From 852bd3e301f688029bdc9782a8a56e6c63660e92 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Date: Wed, 20 Nov 2024 17:18:28 +0100
Subject: [PATCH 744/954] fix(st-crypto): improve RNG health test configuration

The default health test configuration in RNG_HTCR is not good enough.
It implies too many zero random generations.
Set a better value to improve the IP behavior.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 984139e4b92772adb2a8f2860c0c3c66b65acf21 fix(st-crypto): improve RNG health test configuration]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Change-Id: I7e05e52466be68fdbf5439931a9f01dcfa57a24d
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/419862
Tested-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/st/crypto/stm32_rng.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/st/crypto/stm32_rng.c b/drivers/st/crypto/stm32_rng.c
index f917a896e6..be2cb5c544 100644
--- a/drivers/st/crypto/stm32_rng.c
+++ b/drivers/st/crypto/stm32_rng.c
@@ -28,6 +28,9 @@
 #define RNG_CR			0x00U
 #define RNG_SR			0x04U
 #define RNG_DR			0x08U
+#if STM32_RNG_VER == 4
+#define RNG_HTCR		0x10U
+#endif
 
 #define RNG_CR_RNGEN		BIT(2)
 #define RNG_CR_IE		BIT(3)
@@ -52,6 +55,14 @@
 #define RNG_NIST_CONFIG_C	0x00F00D00U
 #define RNG_NIST_CONFIG_MASK	GENMASK(25, 8)
 
+#if STM32_RNG_VER == 4
+#if STM32_RNG_VER_MINOR == 2
+#define RNG_HTCFG_CONFIG	0x000072ACU /* Reset value */
+#else
+#define RNG_HTCFG_CONFIG	0x0000AAC7U
+#endif
+#endif
+
 #define RNG_MAX_NOISE_CLK_FREQ	48000000U
 
 struct stm32_rng_instance {
@@ -125,6 +136,8 @@ static int stm32_rng_enable(void)
 	mmio_clrsetbits_32(stm32_rng.base + RNG_CR, RNG_CR_CLKDIV,
 			   (clock_div << RNG_CR_CLKDIV_SHIFT));
 
+	mmio_write_32(stm32_rng.base + RNG_HTCR, RNG_HTCFG_CONFIG);
+
 	mmio_clrsetbits_32(stm32_rng.base + RNG_CR, RNG_CR_CONDRST, RNG_CR_RNGEN);
 #endif
 	timeout = timeout_init_us(RNG_TIMEOUT_US);
-- 
2.53.0

